article.md

```diff
@@ -0,0 +1,333 @@
+```cairo
+@external
+func add_connection{
+    syscall_ptr : felt*,
+    pedersen_ptr : HashBuiltin*,
+    range_check_ptr,
+    bitwise_ptr : BitwiseBuiltin*
+}(eth_address : felt, starknet_address : felt, domain_hash: Uint256, signature_len: felt, signature: felt*){
+    assert_valid_eth_signature(eth_address, starknet_address, domain_hash, signature_len, signature);
+    save_connected_addresses(eth_address, starknet_address);
+    return ();
+}
+```
+
+The hardest part for us is getting the hash of our message. Once we have that, we can use a built-in function `verify_eth_signature_uint256` that will do all the heavy lifting, dealing with complicated math. 
+So we will have to write the `get_hash` function for ourselves.
+```cairo
+func assert_valid_eth_signature{
+        syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr,
+        bitwise_ptr : BitwiseBuiltin*
+}(eth_address: felt, starknet_address: felt, domain_hash: Uint256, signature_len: felt, signature: felt*){
+    alloc_locals;
+    with_attr error_message(
+        "Invalid signature length. Signature should have exactly 5 elements."
+    ){
+        assert signature_len = 5;
+    }
+
+    let v = signature[0];
+    let r = Uint256(signature[2], signature[1]);
+    let s = Uint256(signature[4], signature[3]);
+    
+    let (local keccak_ptr_start) = alloc();
+    let keccak_ptr = keccak_ptr_start;
+
+    let (hash_uint) = get_hash{keccak_ptr=keccak_ptr}(starknet_address, domain_hash);
+   
+    verify_eth_signature_uint256{keccak_ptr=keccak_ptr}(hash_uint, r, s, v, eth_address);
+
+    finalize_keccak(keccak_ptr_start=keccak_ptr_start, keccak_ptr_end=keccak_ptr);
+
+    return ();
+}
+```
+
+## Signing with an Ethereum wallet
+Let's use Metamask as an example. You just need to call the `signTypedData_v4` function (inputting the correct address in the `myStarkNetAddress` field) and it will return the signature that we need.<br/>
+```js
+const msg = {
+  domain: {
+    name: "Confirm the ownership of this StarkNet address",
+    verifyingContract: '0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF',
+    version: '1',
+  },
+  message: {
+    "starknetAddress": 12356789,
+  },
+  primaryType: 'Payload',
+  types: {
+    domain: [
+      { name: 'name', type: 'string' },
+      { name: 'verifyingContract', type: 'address' },
+      { name: 'version', type: 'string' },
+    ],
+    Payload: [
+      { name: 'starknetAddress', type: 'uint256' },
+    ],
+  },
+};
+const accounts = await window.ethereum.request({method: "eth_requestAccounts"})
+await window.ethereum.request({
+    method: "eth_signTypedData_v4",
+    from: accounts[0],
+    params: [accounts[0], JSON.stringify(msg)]
+})
+```
+After you input this JS code into the console on the MetaMask website it will trigger a signing request (of course normally you wouldn't do it this way but it's just for demonstration's sake).<br/><br/>
```

Comment:
"After you run this... in the browser's console"

Instead of sentence in `()` you can add an additional sentence like "An application would use a similar code to get user's accounts and signature. After that it can send a transaction containing both the Ethereum address and the signature to StarkNet's contract."